Use only atomic operations in Plug.Upload to prepare for concurrent writes

magnetised opened the pull request:
Which allows a request to transfer ownership of an uploaded file to
another process.

This makes it easy to run extended file processing or upload outside of the request-response cycle. The alternative to this we've been using is to copy the file and use the copy for processing, which is less than ideal.

I based the function name on `:ets.give_away/3` which does essentially the same thing.

josevalim commented on lib/plug/upload.ex:
```diff
@@ -59,25 +59,66 @@ defmodule Plug.Upload do
+
+        case :ets.lookup(@table, to_pid) do
+          [{^to_pid, _tmp, paths}] ->
+            true = :ets.update_element(@table, to_pid, {3, [path | paths]})
```
These operations need to happen on the server, otherwise you will have race conditions. :)

magnetised replied:
Doh. Ok. Will rearrange.. Thanks

josevalim replied:
For example, if two transfers to the same process happen at the same time.

josevalim commented:
Thank you @magnetised for the PR! We discussed adding this for LV a couple times, so this is exciting!

My biggest concern right now is that this implementation is not really safe, as it may have race conditions. The reason why `random_file` is safe is because we assume `pid = self()`. However, if we allow giving away, then the giving away code (and all existing code) need to be moved to the GenServer, which can become a bottleneck, unless we reimplement the whole code to rely on atomic operations (for example, using two ets tables, one as sets and another as duplicate_bag).

magnetised commented:
@josevalim right. Tricky. So if we rule out the everything-through-the-server as an obvious bottleneck, do you mean using the `set`-type table to keep the `tmp` dir for a given process (and acts as a test of existance/membership), while the `duplicate_bag` keeps the list of paths for a given process?

We can just add and delete paths in the duplicate_bag without danger of race-conditions. Does that sound about right?

josevalim commented:
Exactly!

It is probably worth refactoring it first to use two tables and then we add the new features. :)

josevalim commented on lib/plug/upload.ex:
```diff
@@ -37,7 +37,8 @@ defmodule Plug.Upload do
         }
 
-  @table __MODULE__
+  @pid_table :"#{__MODULE__}_dir"
+  @path_table :"#{__MODULE__}_path"
```
```suggestion
  @pid_table __MODULE__.Dir
  @path_table __MODULE__.Path
```

josevalim commented on lib/plug/upload.ex:
```diff
@@ -155,22 +156,29 @@ defmodule Plug.Upload do
+      [{pid, _tmp}] ->
+        :ets.delete(@pid_table, pid)
+
+        @path_table
+        |> :ets.lookup(pid)
+        |> Enum.map(&elem(&1, 1))
+        |> delete_paths()
```
```suggestion
        @path_table
        |> :ets.lookup(pid)
        |> Enum.each(&delete_path/1)
```

And then in terminate you can do:

    folder = fn entry, :ok -> delete_path(entry) end

josevalim commented:
Two minor comments and we can merge it. :)

magnetised commented:
@josevalim as suggested, I've dropped the `give_away` function and refactored to use two ets tables.

This is not a huge change but I'm not sure we've solved the problem. 

Going back to the give_away function still requires inserting an entry for the *receiving* pid into the `@pid_table` from the existing owning process (say the request process), rather than the destination process, so we still have a race condition but not such a serious one.

I did consider having two classes of members: those who have created an upload and those who have only received them, but we need to monitor all processes. To avoid duplicate calls to `Process.monitor/1`, which would be messy, we need an entry in the `@pid_table` even if a process has only ever received a file.

Possible solutions:

1. We could have a *third* ets table which only contains a `{pid, true}` and use the result of `:ets.insert_new/2` as an atomic test for existence (followed by the longer running insertion into the ets mapping pids to tmp roots.

2. Ignore the race condition: it just results in a bit of extra work, potentially finding a valid tmp dir before finding that some other process has just done the same thing and `:ets.insert_new` fails (returns `false`)

3. Allow duplicate calls to `Process.monitor` for the same pid, and only add entries to the `@pid_table` when calling `random_file/1`. So a process receiving a file gets monitored and just adds a path to the `@path_table`. In the `{:DOWN, ...}` handler we just ignore messages from pids that we don't have entries for (and always remove paths even if they don't exist in the `@pid_table`.
 
4. We could change the api to our `give_away` to move adopting of the file into the receiving process, which could look something like this:

```elixir
# in request process
gifted_upload = Plug.Upload.give_away(%Plug.Upload{} = upload)
# gifted upload contains a ref to the request process, e.g. {%Plug.Upload{}, pid}
{:ok, pid} = SomeModule.start_link(upload: gifted_upload)

# in receiving process

def init(args) do
  # since we're doing the work from the receiving process, we can safely write to the pid -> tmp root ets
  {:ok, upload} = Plug.Upload.adopt(args[:upload])
  {:ok, upload, {:continue, :process_upload}
end
```

I like 4 because it makes it easy to start a child process and immediately receive ownership of the uploaded file (rather than having to start the process, gift the file then start processing). But it's a bit of faff.

2 is obviously the simplest.

3 is straight-forward and avoids race-conditions - but if you have a long running process processing a lot of files the `Plug.Upload` process may receive a *lot* of `:DOWN` messages when it terminates. Having said that this entire API is designed for short-lived processes, so a single long lived process owning uploads would result in a lot of random tmp files...

What do you think?

josevalim commented:
> Going back to the give_away function still requires inserting an entry for the receiving pid into the @pid_table from the existing owning process (say the request process), rather than the destination process, so we still have a race condition but not such a serious one.

I don't see the issue because inserting into the table is always atomic. The only issue I can think is that the process we are transferring from dies during the transfer and the file is lost but we can solve it by making sure we can only transfer from the owner process.

josevalim commented:
To be clear, I think the following order of operations happening from the current owner process is fine:

1. ask for the receiving process to be monitored and ensure the tmp dir exists for the receiving process
2. write the path under the receiving process
3. remove the path from the current process

WDYT?

magnetised commented:
The only issue I can see, which has made me wonder if there was another way, is this:

> ask for the receiving process to be monitored and **ensure the tmp dir exists for the receiving process**

(emphasis mine) - ensuring the temp dir exists is not atomic, *but* the worst that happens is that two processes simultaneously try to create a tmp dir for the current process and we'll have a first-write-wins situation on the dir_table.

It's just messy, not critical

I'm assuming you're ok with this minor issue. 

I might implement my give-receive process as an alternate API as part of the forthcoming PR because, as I say, I like the ability to start a process and be able to adopt a file as part of the init callback. It's just a wrapper around passing the existing owner pid to the receiving process.

josevalim commented:
> (emphasis mine) - ensuring the temp dir exists is not atomic, but the worst that happens is that two processes simultaneously try to create a tmp dir for the current process and we'll have a first-write-wins situation on the dir_table.

This is easy to fix because of `insert_new`. If `insert_new` returns true, than whatever you gave is the tmp directory. If it returns false, then you can read whatever is stored. 

However, I realize there is another potential bug in that between `GenServer.call({:monitor, pid})` and `insert_new`, the process can crash and the server may clean it up, meaning we will introduce an entry in the table that will never be cleaned. So in the transfer cases, the monitoring and the writing to the table needs to be done in the process atomically. And then you return the result of `insert_new`.

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

josevalim commented:
TL;DR for my comment above: in the transfer case, make the monitor+insert serial on the server.
